docs/src/tutorials/hybrid-applications-with-webpack/template-deep-dive/src-lib-rs.md

```diff
@@ -0,0 +1,306 @@
+code at all, and hence the default memory allocator will be
+used instead of `wee_alloc`.
+
+```rust
+extern crate wee_alloc;
+#[global_allocator]
+static ALLOC: wee_alloc::WeeAlloc = wee_alloc::WeeAlloc::INIT;
+```
+
+However, suppose `"wee_alloc"` is appended to the `default`
+vector in `Cargo.toml`. Then, the `cfg_if!` block is instead
+replaced with the contents of the `if` block, shown above.
+
+This code sets the `wee_alloc` allocator to be used as the
+global memory allocator.
+
+### What is `wee_alloc`?
+
+Reducing the size of compiled WebAssembly code is important,
+since it is often transmitted over the Internet or placed on
+embedded devices.
+
+> `wee_alloc` is a tiny allocator designed for WebAssembly
+> that has a (pre-compression) code-size footprint of only a
+> single kilobyte.
+
+[An analysis](http://fitzgeraldnick.com/2018/02/09/wee-alloc.html)
+suggests that over half of the bare minimum WebAssembly
+memory footprint is required by Rust's default memory
+allocator. Yet, WebAssembly code often does not require a
+sophisticated allocator, since it often just requests a
+couple of large initial allocations.
+
+`wee_alloc` trades off size for speed. Although it has a
+tiny code-size footprint, it is relatively slow if
+additional allocations are needed.
+
+For more details, see the
+[`wee_alloc` repository](https://github.com/rustwasm/wee_alloc).
+
+## 4. Defining `set_panic_hook`
+
+```rust
+cfg_if! {
+    // When the `console_error_panic_hook` feature is enabled, we can call the
+    // `set_panic_hook` function to get better error messages if we ever panic.
+    if #[cfg(feature = "console_error_panic_hook")] {
+        extern crate console_error_panic_hook;
+        use console_error_panic_hook::set_once as set_panic_hook;
+    } else {
+        #[inline]
+        fn set_panic_hook() {}
+    }
+}
+```
+
+As described in the `wee_alloc` section, the macro `cfg_if!`
+evaluates the `if` statement during compile time. This is
+possible because it is essentially testing whether
+`"console_error_panic_hook"` is defined in the `[features]`
+section of `Cargo.toml`, which is available during compile
+time.
+
+The entire macro block will either be replaced with the
+statements in the `if` block or with those in the `else`
+block. These two cases are now described in turn:
+
+```rust
+extern crate console_error_panic_hook;
+use console_error_panic_hook::set_once as set_panic_hook;
+```
+
+Due to the `use` statement, the function
+`console_error_panic_hook::set_once` can now be
+accessed more conveniently as `set_panic_hook`.
+
+```rust
+#[inline]
+fn set_panic_hook() {}
+```
+
+An inline function replaces the function call with the
+contents of the function during compile time. Here,
+`set_panic_hook` is defined to be an empty inline function.
+This allows the use of `set_panic_hook` without any run-time
+or code-size performance penalty if the feature is not
+enabled.
+
+### What is `console_error_panic_hook`?
```

Comment:
This section looks duplicated with the description of `console_error_panic_hook` in the webpack section, so I wonder if they could be deduplicated?